src/main/java/uk/gov/justice/hmpps/offenderevents/subscribe/PrisonerEventsListener.kt

```diff
@@ -0,0 +1,67 @@
+
+package uk.gov.justice.hmpps.offenderevents.subscribe
+
+import com.fasterxml.jackson.core.JsonProcessingException
+import com.fasterxml.jackson.databind.ObjectMapper
+import io.awspring.cloud.sqs.annotation.SqsListener
+import io.awspring.cloud.sqs.listener.QueueAttributes
+import io.opentelemetry.api.trace.SpanKind
+import io.opentelemetry.instrumentation.annotations.WithSpan
+import org.slf4j.LoggerFactory
+import org.springframework.beans.factory.annotation.Value
+import org.springframework.stereotype.Service
+import software.amazon.awssdk.services.sqs.model.SendMessageRequest
+import uk.gov.justice.hmpps.offenderevents.model.OffenderEvent
+import uk.gov.justice.hmpps.offenderevents.services.HMPPSDomainEventsEmitter
+import uk.gov.justice.hmpps.sqs.HmppsQueueService
+import java.time.Duration
+import java.time.OffsetDateTime
+
+@Service
+class PrisonerEventsListener(
+  private val objectMapper: ObjectMapper,
+  private val eventsEmitter: HMPPSDomainEventsEmitter,
+  private val hmppsQueueService: HmppsQueueService,
+  @Value("\${application.listener.totalDelayDuration}") private val totalDelay: Duration,
+  @Value("\${application.listener.delayDuration}") private val delay: Duration,
+) {
+  @SqsListener(queueNames = ["prisoneventqueue"], factory = "hmppsQueueContainerFactoryProxy")
+  @WithSpan(value = "Digital-Prison-Services-dev-prisoner_offender_events_queue", kind = SpanKind.SERVER)
+  @Throws(JsonProcessingException::class)
+  fun onPrisonerEvent(message: String?, attributes: QueueAttributes) {
+    val sqsMessage: SQSMessage = objectMapper.readValue(message, SQSMessage::class.java)
+    val publishedAt = OffsetDateTime.parse(sqsMessage.MessageAttributes.publishedAt.Value)
+    val event = objectMapper.readValue(sqsMessage.Message, OffenderEvent::class.java)
+    if (!DELAYED_EVENT_TYPES.contains(event.eventType) || publishedAt.isBefore(
+        OffsetDateTime.now().minus(totalDelay),
+      )
+    ) {
+      log.debug(
+        "Received message {} type {} published at {}",
+        sqsMessage.MessageId,
+        event.eventType,
+        publishedAt,
+      )
+      eventsEmitter.convertAndSendWhenSignificant(event)
+    } else {
+      hmppsQueueService.findByQueueId("prisoneventqueue")!!.sqsClient
+        .sendMessage(
+          SendMessageRequest.builder()
+            .queueUrl(attributes.queueUrl)
+            .messageBody(message)
+            .delaySeconds(delay.toSeconds().toInt())
+            .build(),
+        )
+    }
+  }
+
+  private companion object {
+    private val DELAYED_EVENT_TYPES = listOf("OFFENDER_MOVEMENT-RECEPTION", "OFFENDER_MOVEMENT-DISCHARGE")
+    private val log = LoggerFactory.getLogger(this::class.java)
+  }
+}
+
+internal data class SQSMessage(val Message: String, val MessageId: String, val MessageAttributes: MessageAttributes)
+internal data class MessageAttributes(val publishedAt: PublishedAt)
+internal data class PublishedAt(val Value: String, val Type: String)
```

Comment:
had to add a `Type` here for some reason, even though not used, in order to get the hydration working. Odd.

Reply:
You could try using the kotlin objectMapper.readValue<class> () method which might help

Reply:
line 33 does
```
    val sqsMessage: SQSMessage = objectMapper.readValue(message, SQSMessage::class.java)
```
anyway, how is that different?  I've checked the object mapper and the kotlin module is registered already, hence the data classes are converted.

Reply:
I meant the readValue(json) method , without the ::class param. afair it does a better job of deserialising nested classes.

Reply:
I've just had a go.  I think what you mean is add in
```
import com.fasterxml.jackson.module.kotlin.readValue
```
and then call
```
    val sqsMessage: SQSMessage = objectMapper.readValue(message!!)
```
instead.  That gives exactly the same error:
```
Caused by: com.fasterxml.jackson.databind.exc.MismatchedInputException: Cannot construct instance of `uk.gov.justice.hmpps.offenderevents.subscribe.PublishedAt` (although at least one Creator exists): cannot deserialize from Object value (no delegate- or property-based Creator)
```
when I remove the `Type` field and just leave it as
```
internal data class PublishedAt(val Value: String)
```